src/trezor-keyring.ts

```diff
@@ -281,44 +351,48 @@ class TrezorKeyring extends EventEmitter {
 
   /**
    *
-   * @template {TypedTransaction | OldEthJsTransaction} Transaction
-   * @param {string} address - Hex string address.
-   * @param {number} chainId - Chain ID
-   * @param {Transaction} tx - Instance of either new-style or old-style ethereumjs transaction.
-   * @param {(import('trezor-connect').EthereumSignedTx) => Transaction} handleSigning - Converts signed transaction
+   * @param address - Hex string address.
+   * @param chainId - Chain ID
+   * @param tx - Instance of either new-style or old-style ethereumjs transaction.
+   * @param handleSigning - Converts signed transaction
    * to the same new-style or old-style ethereumjs-tx.
-   * @returns {Promise<Transaction>} The signed transaction, an instance of either new-style or old-style
+   * @returns The signed transaction, an instance of either new-style or old-style
    * ethereumjs transaction.
    */
-  async _signTransaction(address, chainId, tx, handleSigning) {
-    let transaction;
+  async #signTransaction<T extends TypedTransaction | OldEthJsTransaction>(
```

Comment:
Hmm, we can use `extends` here for tersity, but it's technically incorrect (unless we want to be allowing input that's an extension of those objects?) And we would be returning a less narrow type,

The way to overload this would be:

```ts
  private async #signTransaction(
    address: string,
    chainId: number,
    tx: OldEthJsTransaction,
    handleSigning: (
      tx: EthereumSignedTx
    ) => (OldEthJsTransaction )
  ): Promise<OldEthJsTransaction>;
  private async #signTransaction(
    address: string,
    chainId: number,
    tx: TypedTransaction,
    handleSigning: (
      tx: EthereumSignedTx,
    ) => (TypedTransaction)
  ): Promise<TypedTransaction>;
  private async #signTransaction(
    address: string,
    chainId: number,
    tx: OldEthJsTransaction | TypedTransaction,
    handleSigning: (
      tx: EthereumSignedTx,
    ) => (OldEthJsTransaction | TypedTransaction)
  ): Promise<OldEthJsTransaction | TypedTransaction> {
```

Reply:
`OldEthJsTransaction | TypedTransaction` can be potentially replaced with `any` in the last overload too AFAIK

Reply:
>but it's technically incorrect (unless we want to be allowing input that's an extension of those objects?

Hmm, it sounds like you're misunderstanding this keyword. It really doesn't imply any sort of OOP-style inheritance. Nothing about using extends with a type union is incorrect. This is a conventional and effective way to ensure that a generic parameter is one of two types.

Reply:
>  This is a conventional and effective way to ensure that a generic parameter is one of two types.

Yes, but it allows for a different type that happens to extend `OldEthJsTransaction | TypedTransaction` to be passed in too - which I don't think we want to allow. E.g.:

```ts
type Tx1 = {
    name1: string;
}

type Tx2 = {
    name2: number;
}

const test = <A extends Tx1 | Tx2>(a:A, fn: (x: any) => A): A => {
    return a;
}

const test1 = test({name1: "test", extraKey: "test"}, (x) => ({name1: "test", extraKey: "test"})); // ✅
const test2 = test({name1: "test", name2: "test"}, (x) => ({name1: "test", name2: "test"})); // ✅
```

Whereas if we use overloads:

```ts
type Tx1 = {
    name1: string;
}

type Tx2 = {
    name2: number;
}

function test (a: Tx1, fn: (x: any) => Tx1): Tx1;
function test (a: Tx2, fn: (x: any) => Tx2): Tx2;
function test (a: Tx1 | Tx2, fn: (x: any) => Tx1 | Tx2): Tx1 | Tx2 {
    return a;
}

const test1 = test({name1: "test", extraKey: "test"}, (x) => ({name1: "test", extraKey: "test"})); // ❌
const test2 = test({name1: "test", name2: "test"}, (x) => ({name1: "test", name2: "test"})); // ❌
const test3 = test({name1: "test"}, (x) => ({name1: "test"})); // ✅
const test4 = test({name2: 4}, (x) => ({name2: 4})); // ✅
```

Reply:
Ahh I see. Interesting, good point.

Reply:
Using `any` in the implementation does effectively disable type checking though, which isn't ideal. We should at least use the type union there, getting us at least as much type safety as we had before.

Thinking about this further though, this seems based on the false assumption that we're expecting exact transaction objects here. The keyring interface we're using doesn't demand that, it allows additional properties. That is, the old type union seems more-correct even with the surprising behavior you've demonstrated in this code snippet.

Reply:
@Gudahtt I had a look into the code today. Given that the tx objects are being created within the keyring itself, and not publically, is there a scenario where we would want to allow the extension of the types?

I've found another benefit to using overloads here, as old transactions should be passing through a Buffer instead of a number. (See my comment on the overload)

Reply:
@andrewpeters9 [see my comment](https://github.com/MetaMask/eth-trezor-keyring/pull/161#discussion_r1127649184) regarding the chainId coming from the old tx.
Moreover, recently a `Keyring` type has been added to `utils` (See https://github.com/MetaMask/utils/blob/main/src/keyring.ts#L137) so we should refactor this package to use that interface in a separate PR, which uses just the `TypedTransaction` - this should simplify a lot the implementation too as we will not expect an old tx anymore. 

Do you think we could maybe track the object extensibility problem you are highlighting in a separate issue?

Reply:
>Given that the tx objects are being created within the keyring itself, and not publically

They're currently being created in the transaction controllers (in the core repo for mobile, and in metamask-extension for the extension).